src/replit_river/v2/session.py

```diff
@@ -0,0 +1,1280 @@
+from aiochannel.errors import ChannelClosed
+from opentelemetry.trace import Span, use_span
+from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator
+from pydantic import ValidationError
+from websockets.asyncio.client import ClientConnection
+from websockets.exceptions import ConnectionClosed, ConnectionClosedOK
+
+from replit_river.common_session import (
+    ConnectingStates,
+    SendMessage,
+    SessionState,
+    TerminalStates,
+    buffered_message_sender,
+)
+from replit_river.error_schema import (
+    ERROR_CODE_CANCEL,
+    ERROR_CODE_SESSION_STATE_MISMATCH,
+    ERROR_CODE_STREAM_CLOSED,
+    ERROR_HANDSHAKE,
+    RiverError,
+    RiverException,
+    RiverServiceException,
+    SessionClosedRiverServiceException,
+    StreamClosedRiverServiceException,
+    exception_from_message,
+)
+from replit_river.messages import (
+    FailedSendingMessageException,
+    WebsocketClosedException,
+    parse_transport_msg,
+    send_transport_message,
+)
+from replit_river.rate_limiter import LeakyBucketRateLimit
+from replit_river.rpc import (
+    ACK_BIT,
+    STREAM_OPEN_BIT,
+    ControlMessageHandshakeRequest,
+    ControlMessageHandshakeResponse,
+    ExpectedSessionState,
+    TransportMessage,
+    TransportMessageTracingSetter,
+)
+from replit_river.seq_manager import (
+    InvalidMessageException,
+    OutOfOrderMessageException,
+)
+from replit_river.task_manager import BackgroundTaskManager
+from replit_river.transport_options import (
+    MAX_MESSAGE_BUFFER_SIZE,
+    TransportOptions,
+    UriAndMetadata,
+)
+
+PROTOCOL_VERSION = "v2.0"
+
+STREAM_CANCEL_BIT_TYPE = Literal[0b00100]
+STREAM_CANCEL_BIT: STREAM_CANCEL_BIT_TYPE = 0b00100
+STREAM_CLOSED_BIT_TYPE = Literal[0b01000]
+STREAM_CLOSED_BIT: STREAM_CLOSED_BIT_TYPE = 0b01000
+
+
+logger = logging.getLogger(__name__)
+
+trace_propagator = TraceContextTextMapPropagator()
+trace_setter = TransportMessageTracingSetter()
+
+CloseSessionCallback: TypeAlias = Callable[["Session"], Coroutine[Any, Any, Any]]
+RetryConnectionCallback: TypeAlias = Callable[
+    [],
+    Coroutine[Any, Any, Any],
+]
+
+
+@dataclass
+class _IgnoreMessage:
+    pass
+
+
+class Session:
+    _transport_id: str
+    _to_id: str
+    session_id: str
+    _transport_options: TransportOptions
+
+    # session state, only modified during closing
+    _state: SessionState
+    _close_session_callback: CloseSessionCallback
+    _close_session_after_time_secs: float | None
+    _connecting_task: asyncio.Task[Literal[True]] | None
+    _wait_for_connected: asyncio.Event
+
+    # ws state
+    _ws: ClientConnection | None
+    _heartbeat_misses: int
+    _retry_connection_callback: RetryConnectionCallback | None
+
+    # message state
+    _process_messages: asyncio.Event
+    _space_available: asyncio.Event
+
+    # stream for tasks
+    _streams: dict[str, Channel[Any]]
+
+    # book keeping
+    _ack_buffer: deque[TransportMessage]
+    _send_buffer: deque[TransportMessage]
```

Comment:
I assume we're using a deque here because .pop(0) reshifts the underlying array?

Reply:
dequeue because append and popleft are both O(1), `list` is O(n) for shifting the whole buffer.

List would be fast enough, but it's still wasteful when there's a stdlib collection that does what we want.